### ISSUE TITLE
key vs endkey

### ISSUE THREAD

**Issue by ronag** [apache/couchdb#3977]
I believe that if `key` is provided `endkey` should be ignored? The docs seem to be unclear on this. However, the behavior I get is the opposite.

```
await couch.put('a', null, {})
await couch.put('b', null, {})
await couch.put('c', null, {})

const { rows } = await couch.allDocs({
  key: 'a',
  endkey: 'b'
})
console.log(rows)
```

```
test/index.js 2> [
test/index.js 2>   {
test/index.js 2>     id: 'a',
test/index.js 2>     key: 'a',
test/index.js 2>     value: { rev: '1-967a00dff5e02add41819138abb3284d' }
test/index.js 2>   },
test/index.js 2>   {
test/index.js 2>     id: 'b',
test/index.js 2>     key: 'b',
test/index.js 2>     value: { rev: '1-967a00dff5e02add41819138abb3284d' }
test/index.js 2>   }
test/index.js 2> ]
```

**Comment by nickva** [comment]
@ronag thanks for your report.

`key` effectively behaves as if we are setting both `startkey=$key&endkey=$key`. So every time it is parsed it does exactly that.

However, we do have "keys" which returns an error as you mentioned:


```
http $DB/db/_all_docs'?keys=["c"]&endkey="b"'
HTTP/1.1 400 Bad Request
{
    "error": "query_parse_error",
    "reason": "`keys` is incompatible with `key`, `start_key` and `end_key`"
}
```

Perhaps `key` it should be translated to `keys=[Key]` internally, instead and behave similarly, but it maybe a compatibility issue in some cases as well. It could be a nice improvement for the future.

**Comment by ronag** [comment]
@nickva I don't think your statement is consistent with how couch behaves. See my example. If you `key` is the same as `startkey=key&endkey=key` then the response should only be `a`. However, that is not the case.

**Comment by ronag** [comment]
@nickva What is the predence here though? It seems setting both key and endkey has different result that just setting key. Or does the order of the parameters matter?

**Comment by big-r81** [comment]
Hi @ronag,

I would say that your example is working like expected, but you are right it seems a little bit confusing and needs an optimization in the future. 
You are accesing the `_all_docs` [view](https://docs.couchdb.org/en/stable/api/database/bulk-api.html#db-all-docs) endpoint and some [view parameters](https://docs.couchdb.org/en/stable/api/ddoc/views.html#db-design-design-doc-view-view-name) are setting default values like `inclusive_end`, which is set to true. 

So you would also get `b`in your result list. Maybe you can test it with `?inclusive_end=false` and if `b`isn't in your result, then it is this default parameter...

**Comment by ronag** [comment]
> The precedence is in the order the parameters are specifie

This answered my confusion.

**Comment by nickva** [comment]
@abarbosa-godaddy yeah, that's the idea of the tag. To do some validation that some parameters are mutually exclusive.
